package.json

```diff
@@ -0,0 +1,36 @@
+{
+  "name": "json-schema-org/json-schema-spec",
+  "version": "1.0.0",
+  "description": "The JSON Schema Specification",
+  "type": "module",
+  "main": "index.js",
+  "scripts": {
+    "lint": "eslint build/",
+    "build": "npm run build-core && npm run build-validation",
+    "build-core": "node build/build.js < jsonschema-core.md > jsonschema-core.html",
+    "build-validation": "node build/build.js < jsonschema-validation.md > jsonschema-validation.html"
+  },
+  "license": "MIT",
+  "dependencies": {
+    "mdast-builder": "^1.1.1",
+    "mdast-util-find-and-replace": "^3.0.0",
+    "mdast-util-to-string": "^4.0.0",
+    "rehype-stringify": "^9.0.3",
+    "remark": "^14.0.3",
+    "remark-flexible-containers": "^1.0.6",
+    "remark-gfm": "^3.0.1",
+    "remark-heading-id": "^1.0.0",
+    "remark-preset-lint-markdown-style-guide": "^5.1.3",
+    "remark-rehype": "^10.1.0",
+    "remark-torchlight": "^0.0.5",
+    "remark-validate-links": "^12.1.1",
+    "unist-builder": "^4.0.0",
+    "vfile-reporter": "^8.0.0"
+  },
+  "devDependencies": {
+    "dotenv": "*",
+    "eslint": "*",
+    "eslint-import-resolver-node": "*",
+    "eslint-plugin-import": "*"
```

Comment:
I've come to believe that pinning versions is a bad practice, especially for dev dependencies. It gets you stuck on out-dated and often abandoned and insecure dependencies. I want my dependencies to be evergreen. Not pinning versions does mean that things will break from time to time, but that's a good thing. When breaking changes are introduced, I want my project to break so I have to deal with them immediately rather than allowing those changes to pile up making it much harder to deal with later. 

Our current blog site is a perfect example of this problem. Its dependencies are many years out of date. Many are no longer supported and have unresolved security vulnerabilities. I've tried and failed a couple of times to update the dependencies with the intention of resolving as many security vulnerabilities as I can. Fixing one thing breaks another and so on until I'm blocked by something or the changes just get to be too much. If the dependencies had been updated to the latest as they changed, it would have probably been relatively easy to make the necessary changes because they would have come in one at a time and have been relatively small and not entangled with other changes.

The approach I've been using on my projects for the last few years is to use `*` for dev dependencies and versions with the `^` prefix for dependencies. I'm really happy with how this has been working out. My builds rarely break and when they do, it's always been a simple fix. I always have the latest features available to me and security vulnerabilities are always quickly and easily resolved just by updating my dependencies.

Let me know if you're willing to give this approach a try. If not, I'll add some versions.

Reply:
I think using major version fix (`^`) makes senes.
I'm VERY aprehensive about not pinning (or at least major version pin) dev dependencies.

> If the dependencies had been updated to the latest as they changed...

I've personally encountered a situation where updating a depencency was impossible, after multiple days of trying to figure out why. Turned out, they released an update a few days after we decided to stop trying to fix a compatibility bug. We would do this activity on a monthly basis. Our test coverage was OK.

I expect at some point I will suggest we use Prettier. They continue to strongly recommend exact version pinning, because updated versions could result in different code formatting, which would fail validation checking done in a remote action/build step.

> Not pinning versions does mean that things will break from time to time, but that's a good thing.

I disagree. A PR which adds a new article or fixes a typo shouldn't fail to build because one of the dev dependencies has made a new release. Say several dev dependencies are updated at once. We now do not know which is causing the break, and we don't know which version was "good".

Many of the security related issues for the website have only been a concern for sites which have a backend or user created content, so have been ignored. This will still be the case, but we can be better at updating dependencies when they have security issues.

Reply:
> I'm VERY aprehensive about not pinning (or at least major version pin) dev dependencies.

Ok. I'll add major version pinning. I think it creates a risk for the long-term health of the build, but it's a sufficiently small project that it shouldn't be a significant issue.

> A PR which adds a new article or fixes a typo shouldn't fail to build because one of the dev dependencies has made a new release.

This wouldn't happen. Dependencies don't get updated by surprise. We have to update them and they get pinned by package-lock.json (thank you for catching my mistake there). If an update breaks something, fix the typo first, then address the dependency issue.

> Say several dev dependencies are updated at once.

That kind of thing would be statistically unlikely if you're updating regularly. (Dependabot can help with that regularity.) When updates happen regularly, breaking changes are rare and more than one package being involved in a breaking change is so rare that I've never seen it in the years I've been working this way (continuous integration). That only happens when trying to update dependencies that haven't been updated regularly for years. Yes, it can happen, but even if there's a break that's difficult to resolve, it's still the best thing to resolve it asap. If you pin it and ignore it, the problem doesn't go away, it just gets harder to resolve later as more dependencies change. Then it doesn't end up getting fixed and you end up with mountains of technical debt that can't be fixed just like our blog site. If you have a dependency that's regularly causing problems, you're better off considering alternatives for that dependency rather than allowing it to bring down the health of your application.